libvast/vast/format/json.hpp

```diff
@@ -41,4 +53,202 @@ class writer : public printer_writer<event_printer>{
+  /// @param input The stream of logs to read.
+  explicit reader(caf::atom_value table_slice_type,
+                  std::unique_ptr<std::istream> in = nullptr);
+
+  void reset(std::unique_ptr<std::istream> in);
+
+  caf::error schema(vast::schema sch) override;
+
+  caf::error schema(vast::type, vast::schema);
+
+  vast::schema schema() const override;
+
+  const char* name() const override;
+
+protected:
+  caf::error read_impl(size_t max_events, size_t max_slice_size,
+                       consumer& f) override;
+
+private:
+  using iterator_type = std::string_view::const_iterator;
+
+  // void patch(std::vector<data>& xs);
+
+  Selector selector;
+  std::unique_ptr<std::istream> input_;
+  std::unique_ptr<vast::detail::line_range> lines_;
+  caf::optional<size_t> proto_field_;
+  std::vector<size_t> port_fields_;
+};
+
+template <class Selector>
+reader<Selector>::reader(caf::atom_value table_slice_type,
+                         std::unique_ptr<std::istream> in)
+  : super(table_slice_type) {
+  if (in != nullptr)
+    reset(std::move(in));
+}
+
+template <class Selector>
+void reader<Selector>::reset(std::unique_ptr<std::istream> in) {
+  VAST_ASSERT(in != nullptr);
+  input_ = std::move(in);
+  lines_ = std::make_unique<vast::detail::line_range>(*input_);
+}
+
+template <class Selector>
+caf::error reader<Selector>::schema(vast::schema s) {
+  return selector.schema(std::move(s));
+}
+
+template <class Selector>
+vast::schema reader<Selector>::schema() const {
+  return selector.schema();
+}
+
+template <class Selector>
+const char* reader<Selector>::name() const {
+  return Selector::name();
+}
+
+template <class F>
+struct convert {
+  convert(F&& f) : f_{std::forward<F>(f)} {
+  }
+
+  bool operator()(vast::json::number n, const vast::count_type&) const {
+    auto c = vast::count(n);
+    return f_(std::move(c));
+  }
+  bool operator()(vast::json::number n, const vast::timespan_type&) const {
+    std::chrono::duration<vast::json::number> x{n};
+    auto t = std::chrono::duration_cast<vast::timespan>(x);
+    return f_(std::move(t));
+  }
+  bool operator()(vast::json::string s, const vast::timestamp_type&) const {
+    vast::timestamp t;
+    if (!vast::parsers::timestamp(s, t))
+      return false;
+    return f_(std::move(t));
+  }
+  bool operator()(vast::json::boolean b, const vast::boolean_type&) const {
+    return f_(std::move(b));
+  }
+  bool operator()(vast::json::number n, const vast::port_type&) const {
+    auto p = vast::port(n);
+    return f_(std::move(p));
+  }
+  bool operator()(vast::json::string s, const vast::address_type&) const {
+    vast::address a;
+    if (!vast::parsers::addr(s, a))
+      return false;
+    return f_(std::move(a));
+  }
+  bool operator()(vast::json::array a, const vast::vector_type& v) const {
+    vast::vector d;
+    auto run = [&](auto value) {
+      d.push_back(std::move(value));
+      return true;
+    };
+    auto c = convert<decltype(run)>{std::move(run)};
```

Comment:
Would `convert{run}` do the trick as well? A lambda doesn't need to be moved and C++17 class template deduction probably obviates the need for `decltype(run)`.

Reply:
The type parameter must be passed explicitly because this code is first instantiated with the lamba from `read_impl()`. It tries to create the same type from the current instantiation and does not do any deduction.
The move can be omitted if I change the constructor to take `F` by value, but then we get a copy of the closure. I'd prefer not to have it.